tests/beta/interfaces/test_tape_autograd.py

```diff
@@ -732,3 +730,559 @@ def cost(x, device):
+                ],
+            ]
+        )
+
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_ragged_differentiation(self, dev_name, diff_method, monkeypatch, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        dev = qml.device(dev_name, wires=2)
+        x = np.array(0.543, requires_grad=True)
+        y = np.array(-0.654, requires_grad=True)
+
+        def _asarray(args, dtype=np.float64):
+            if len({i.shape[0] if i.shape else 0 for i in args}):
+                res = [np.reshape(i, [-1]) for i in args]
+                return np.hstack(res)
+            return np.tensor(args, dtype=dtype)
+
+        if dev_name == "default.qubit.autograd":
+            # we need to patch the asarray method on the device
+            monkeypatch.setattr(dev, "_asarray", _asarray)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit(x, y):
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return [expval(qml.PauliZ(0)), probs(wires=[1])]
+
+        res = circuit(x, y)
+
+        expected = np.array(
+            [np.cos(x), (1 + np.cos(x) * np.cos(y)) / 2, (1 - np.cos(x) * np.cos(y)) / 2]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = qml.jacobian(circuit)(x, y)
+        expected = np.array(
+            [
+                [-np.sin(x), 0],
+                [-np.sin(x) * np.cos(y) / 2, -np.cos(x) * np.sin(y) / 2],
+                [np.cos(y) * np.sin(x) / 2, np.cos(x) * np.sin(y) / 2],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_sampling(self, dev_name, diff_method):
+        """Test sampling works as expected"""
+        dev = qml.device(dev_name, wires=2, shots=10)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd")
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return [sample(qml.PauliZ(0)), sample(qml.PauliX(1))]
+
+        res = circuit()
+
+        assert res.shape == (2, 10)
+        assert isinstance(res, np.ndarray)
+
+
+def qtransform(qnode, a, interface=np):
+    """transforms every RY(y) gate in a circuit to RX(-a*cos(y))"""
+
+    def construct(self, args, kwargs):
+        """New quantum tape construct method, that performs
+        the transform on the tape in a define-by-run manner"""
+        global t_op
+
+        QNode.construct(self, args, kwargs)
+
+        new_ops = []
+        for o in self.qtape.operations:
+            if isinstance(o, qml.RY):
+                t_op = qml.RX(-a * interface.cos(o.data[0]), wires=o.wires)
+                new_ops.append(t_op)
+            else:
+                new_ops.append(o)
+
+        self.qtape._ops = new_ops
+        self.qtape._update()
+
+    import copy

```

Comment:
Recall imports like these came up before, but still got me wondering: won't this import be done on every call? :thinking:

Reply:
Nope! Python only imports the first time. If the module has been previously imported, it is simply skipped :)

This is why there aren't any issues with having `import numpy` in multiple modules of a package.